trie: cache collapsed tries node, not rlp blobs

karalabe opened the pull request:
The current trie memory database/cache that we do pruning on stores trie nodes as binary rlp encoded blobs, and also stores the node relationships/references for GC purposes. However, most of the trie nodes (everything apart from a value node) is in essence just a collection of references.

This PR switches out the RLP encoded trie blobs with the collapsed-but-not-serialized trie nodes. This permits most of the references to be recovered from within the node data structure, avoiding the need to track them a second time (expensive memory wise).

githuboy commented:
geth syncing is alaways 70~100 blocks slower than the latest.And the eth.blockNumber is 0. how to fix it?

fjl commented on trie/database.go:
```diff
@@ -85,22 +86,109 @@ type Database struct {
 type cachedNode struct {
-	blob     []byte              // Cached data block of the trie node
+	node node   // Cached collapsed trie node, excludes raw
+	raw  []byte // Cached raw data (code), excludes node
```
You can make a `rawNode` type and store it in `node` instead.

karalabe replied:
We only use this for storing bytecode. No need to cast it to a node.

fjl replied:
What I meant is that you can avoid having this field and save 16 bytes per cachedNode if you store the raw data in the node field instead.

karalabe replied:
Ah, I see. Hmm, I'm unsure how that would pan out when we need to differentiate between real nodes and rlp blobs. Might be doable with a type case though.

karalabe replied:
Done. PTAL

fjl commented on trie/database.go:
```diff
@@ -85,22 +86,109 @@ type Database struct {
+
+	case valueNode:
+		// Values need to convert from "" to nil
+		if n == nil {
```
This is a noop, `return n` should be enough.

karalabe replied:
It is not a noop, `n` is `valueNode(nil)`, which is different from `nil` (the former can be case to a `valueNode`, the latter cannot. Yay Go quirks.

PS: replacing this with `return n` will panic the trie tests.

fjl commented on trie/database.go:
```diff
@@ -85,22 +86,109 @@ type Database struct {
+
+// expandNode traverses the node hierarchy of a collapsed storage node and converts
+// all fields and keys into expanded memory form.
+func expandNode(n node, cachgen uint16) node {
```
`cachegen`

fjl commented on trie/trie.go:
```diff
@@ -434,6 +434,9 @@ func (t *Trie) resolveHash(n hashNode, prefix []byte) (node, error) {
+	if node := t.db.node(hash, t.cachegen); node != nil {
+		return node, nil
+	}
 	enc, err := t.db.Node(hash)
```
You can call `t.db.diskdb.Get(hash)` here to avoid taking the DB lock twice.

karalabe replied:
t.db.Node can pull up RLP blobs from the cache, some that might not yet have been written to the database.

fjl replied:
But you are calling `t.db.node` just above, that also pulls from the cache right?

karalabe replied:
It doesn't do parsing, but I agree it's a bit convoluted, I'll trink about solving this nicer.

karalabe replied:
Done. PTAL

karalabe commented:
Apparently there's no immediate benefit of this PR, at least not according to datadog. We can sit on this a bit more to see where the changes will lead.

karalabe commented:
The original implementation of this PR turned out to not really solve the initial issue it set out to solve. Memory consumption did not go down, rather it went a bit up.

![screenshot from 2018-06-18 10-35-05](https://user-images.githubusercontent.com/129561/41528946-3c699bb8-72f4-11e8-90a0-367abe7926b5.png)

Blue is master. First half purple is master + 50/50% db/trie cache allocations (note, it's faster, but surfaces memory issues in the trie memcache). Second half purple is collapsed trie nodes + 50/50 allocation. Here the allocs become larger, even though in theory I have less data in the cache.

Fun fact, mallocs and gc goes down, so the PR is more optimal from a GC perspective:

![screenshot from 2018-06-18 10-40-43](https://user-images.githubusercontent.com/129561/41529001-627c5ac0-72f4-11e8-84d5-ece61d1a6ea9.png)

---

My current hunches are:

 * Actual trie nodes (short and full) have an embedded cache (hash + some cachegen fields)... these are useless infos in the memcache, but they probably add a lot of weight.
 * I also have a hunch that we have a significant memory/GC overhead due to the collapsed nodes using live `valueNode(nil)` objects instead of nil for terminators.

I've updated the PR to work around these two potential issues. Nodes are never expanded with `valueNode(nil)` from now, rather only a temporary child-list assembled during RLP encoding. This should avoid a `12 byte / node` overhead for `nil` children.

Furthermore I changed the PR so short and full nodes are replaced with stripped down versions that do not contain the `flags` field. Of course there's an overhead of converting collapsed nodes to these simplified ones on insertion and expanding them back during retrieval.

Lets wait and see how these improvements change the benchmarks, if at all.

karalabe commented:
Mallocs are a bit better overall with this PR.

![screenshot from 2018-06-20 11-46-02](https://user-images.githubusercontent.com/129561/41647676-a57595da-747f-11e8-8d50-c0f946ee4e41.png)

Trie memcache pruning time goes up a bit however, my hunch is because we're handling more complex objects + a lot of interface casting. Not sure if there's much we can do about this. For my full pruning experiments to follow, I kind of need the positional infos too.

![screenshot from 2018-06-20 11-47-26](https://user-images.githubusercontent.com/129561/41647740-c5cf8bce-747f-11e8-9815-683546348ef8.png)
